### PR TITLE
Extend the classify command to perform retriggers on unknown tasks (Fixes #534)

### LINKED ISSUE #534: Extend the classify command to perform retriggers and backfills when it needs more information
The algorithm behind the classify command isn't always able to give a decision between "good" or "bad" and sometimes needs more information in order to be sure.

When a situation like this happens and the `classify` command returns "unknown", it should automatically identify and perform the set of actions (retriggers and/or backfills) that would be necessary to gather enough information.

Depends on https://github.com/mozilla/mozci/issues/504 and https://github.com/mozilla/mozci/issues/508.

### PR DESCRIPTION
**Pull request by chichi012**
Fixes #534 

Assumption:
- `regressions.unknown` returns a map of the manifest name mapped to a list of unknown tasks within that manifest.

### CODE REVIEW

**Comment by chichi012** [comment]
@jmaher @ahal Kindly help review when you can. I plan to write tests for my changes when cleared.

File: mozci/console/commands/push.py
Comment (jmaher) [inline mozci/console/commands/push.py:90]: is retrigger a boolean, it probably doesn't need the `=` after it.

File: mozci/push.py
Comment (jmaher) [inline mozci/push.py:1180]: this most likely retriggers the task a single time, typically we will need more data, I would vote for 4 times (when added to the original task, that would equal 5 total tasks to determine status).  The problem is that if we already have a retrigger and both failed for different reasons, we would have a large number of extra tasks that are not needed.

Think about how to extend this, but for the purposes of this review and PR, lets focus on a single retrigger.

**Review by jmaher — CHANGES_REQUESTED** [review]
only 1 issue I see with the `=` after `--retrigger=`

File: mozci/push.py
Comment (chichi012) [reply mozci/push.py:1180]: @jmaher I can extend the code to call retrigger x4 if that is our end goal. Do we want to get that in now?

Comment (jmaher) [reply mozci/push.py:1180]: can you make it a variable for number of retriggers, and then we can set it now to a default value of 1.

Comment (chichi012) [reply mozci/push.py]: Sounds good. Updated the PR

**Comment by jmaher** [comment]
I like the changes, I need to dig in to the test to see if there is reasonable coverage- most likely yes.

**Review by jmaher — CHANGES_REQUESTED** [review]
is there a way to add any tests to push classification for this?
https://github.com/mozilla/mozci/blob/c130bb7055d77b9b551c348f9d85f535ee169e39/tests/test_push.py#L1061

I think ensuring a known set of tasks that are unknown are retriggered via expected.  Also making sure with retrigger=False (default) that no tasks are retriggered.

Outside of that I don't think there is anything needed to land this.

**Comment by chichi012** [comment]
@jmaher @ahal for the test in `test_push.py`, I don't know why `responses.calls` is empty even when I know that HTTP requests were made.

**Comment by chichi012** [comment]
@jmaher @ahal I believe it's ready now. I've been able to resolve the issues. The problem was the different test cases I was using. It tried checking if a retrigger or classification happened even when I didn’t put any tasks in them except one. But it has been rectified now.

File: mozci/console/commands/push.py
Comment (ahal) [inline mozci/console/commands/push.py:90]: Marco and I had chatted about whether or not this would be better as a cli flag, or as a brand new command. I think I'd mildly prefer a new command, but if Marco / Bastien are ok with the flag here, then that's fine with me too.

If we do go the new command route, I'd call it something like `retrigger-unclassified`, and it would basically just do the exact same thing as this current patch does, except it wouldn't dump the classification results to stdout.

**Review by ahal — COMMENTED** [review]
Awesome, this looks great to me!

Since this intersects with the auto classification project, I'm flagging Marco and Bastien for review too.

File: mozci/console/commands/push.py
Comment (chichi012) [reply mozci/console/commands/push.py:90]: Oh, this sounds great @ahal. Do we do it now, or merge and refactor the code later, or wait for Marco/Bastien's review?

**Comment by jmaher** [comment]
wait for a review from either marco or bastien.

File: mozci/console/commands/push.py
Comment (marco-c) [reply mozci/console/commands/push.py:90]: I don't have a preference between the two options (flag vs new command), either wfm.

File: mozci/console/commands/push.py
Comment (marco-c) [inline mozci/console/commands/push.py:130]: The names of the arguments changed, can you rebase?

File: mozci/push.py
Comment (marco-c) [inline mozci/push.py:1157]: Even if we implement the retriggering as a flag of the classify command, I'd write two separate functions, `classify` and `retrigger`.
`retrigger` could accept a list (or iterable) of tasks as input, and the caller will first call `classify` and then call `retrigger` with the list (or iterable) of unknown tasks.

File: tests/test_push.py
Comment (marco-c) [inline tests/test_push.py:1365]: Instead of just checking that we retrigger, it might be worth checking that we retrigger the right things

File: mozci/push.py
Comment (marco-c) [inline mozci/push.py:1177]: This should be an option

Comment (marco-c) [reply mozci/push.py:1177]: I mean, a parameter of the function

File: tests/test_task.py
Comment (marco-c) [inline tests/test_task.py:80]: Nit: why is this needed?

**Comment by marco-c** [comment]
Note: retriggering might not always be the right action, depending on why the task is in the list of unknown. For example, another potential action is "backfilling".
We should choose the best action to perform depending on the reason why a task is in the unknown list.

For now this is OK though, we can implement the "best action" decision in a follow-up.

**Comment by ahal** [comment]
> Note: retriggering might not always be the right action, depending on why the task is in the list of unknown. For example, another potential action is "backfilling". We should choose the best action to perform depending on the reason why a task is in the unknown list.
> 
> For now this is OK though, we can implement the "best action" decision in a follow-up.

Agreed, this sounds like a good follow-up though.

File: tests/test_task.py
Comment (chichi012) [reply tests/test_task.py:80]: We needed it because we log the label of the task that is being retriggered.

File: mozci/console/commands/push.py
Comment (chichi012) [reply mozci/console/commands/push.py]: okay done

File: mozci/push.py
Comment (chichi012) [reply mozci/push.py]: okay

File: mozci/push.py
Comment (chichi012) [reply mozci/push.py]: okay noted.

File: mozci/push.py
Comment (marco-c) [inline mozci/push.py:1185]: Nit: keep previous wording

Comment (chichi012) [reply mozci/push.py]: okay

File: mozci/console/commands/push.py
Comment (jmaher) [reply mozci/console/commands/push.py]: @chichi012 is this conversation ready to resolve?

File: mozci/push.py
Comment (marco-c) [inline mozci/push.py:1329]: Nit: one thing I didn't notice before, you should add type definitions to all arguments and a return type:
```suggestion
def retrigger(tasks: List[TestTask], repeat_retrigger: int=1) -> None:
```

otherwise mypy won't perform type checking for this function and its calls.

Comment (chichi012) [reply mozci/push.py]: Okay noted. Fixed

**Pull request merged**
